From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sebastian Bresin <sebastian.bresin@gmail.com>
Date: Tue, 7 Apr 2026 12:00:00 +0200
Subject: [PATCH 1/2] feat: add 'leadingspacewidth' buffer option (lsw)

Add a new buffer-local option that controls the visual width of leading
space characters without modifying buffer content. When set to N > 0,
each leading space renders as N screen columns wide. Default is 0
(disabled). Abbreviation: lsw.

This is useful when working with codebases that use narrow space
indentation (e.g. 2-space) but you prefer a wider visual indent.
With 2-space indentation and lsw=2, each indent level displays as
4 columns wide.

Ported for Neovim 12+:
- optwin integration now patches runtime/scripts/optwin.lua
- drawline uses a dedicated contiguous leading-space boundary
- includes updated unit and UI tests

Ref: neovim#15353
---
diff --git a/runtime/scripts/optwin.lua b/runtime/scripts/optwin.lua
index f91cbedc67..a406e5df09 100644
--- a/runtime/scripts/optwin.lua
+++ b/runtime/scripts/optwin.lua
@@ -85,6 +85,7 @@ local options_list = {
     { 'redrawdebug', N_ 'change the way redrawing works (debug)' },
     { 'list', N_ 'show <Tab> as ^I and end-of-line as $' },
     { 'listchars', N_ 'list of strings used for list mode' },
+    { 'leadingspacewidth', N_ 'visual width of leading space characters' },
     { 'number', N_ 'show the line number for each line' },
     { 'relativenumber', N_ 'show the relative line number for each line' },
     { 'numberwidth', N_ 'number of columns to use for the line number' },
diff --git a/src/nvim/buffer_defs.h b/src/nvim/buffer_defs.h
index 6b1719cc6e..e09565911f 100644
--- a/src/nvim/buffer_defs.h
+++ b/src/nvim/buffer_defs.h
@@ -599,6 +599,7 @@ struct file_buffer {
   int b_p_pi;                   ///< 'preserveindent'
   char *b_p_qe;                 ///< 'quoteescape'
   int b_p_ro;                   ///< 'readonly'
+  OptInt b_p_lsw;               ///< 'leadingspacewidth'
   OptInt b_p_sw;                ///< 'shiftwidth'
   OptInt b_p_scbk;              ///< 'scrollback'
   int b_p_si;                   ///< 'smartindent'
diff --git a/src/nvim/change.c b/src/nvim/change.c
index b5533f273f..c14db3bb67 100644
--- a/src/nvim/change.c
+++ b/src/nvim/change.c
@@ -1068,7 +1068,7 @@ bool open_line(int dir, int flags, int second_line_indent, bool *did_do_comment)
     // If 'autoindent' and/or 'smartindent' is set, try to figure out what
     // indent to use for the new line.
     // count white space on current line
-    newindent = indent_size_ts(saved_line, curbuf->b_p_ts, curbuf->b_p_vts_array);
+    newindent = indent_size_ts(saved_line, curbuf->b_p_ts, curbuf->b_p_vts_array, 0);
     if (newindent == 0 && !(flags & OPENLINE_COM_LIST)) {
       newindent = second_line_indent;  // for ^^D command in insert mode
     }
@@ -1512,7 +1512,7 @@ bool open_line(int dir, int flags, int second_line_indent, bool *did_do_comment)
 
         // Recompute the indent, it may have changed.
         if (curbuf->b_p_ai || do_si) {
-          newindent = indent_size_ts(leader, curbuf->b_p_ts, curbuf->b_p_vts_array);
+          newindent = indent_size_ts(leader, curbuf->b_p_ts, curbuf->b_p_vts_array, 0);
         }
 
         // Add the indent offset
diff --git a/src/nvim/drawline.c b/src/nvim/drawline.c
index d2f9f17cc0..30feefff37 100644
--- a/src/nvim/drawline.c
+++ b/src/nvim/drawline.c
@@ -1438,6 +1438,7 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, b
 
   colnr_T trailcol = MAXCOL;  // start of trailing spaces
   colnr_T leadcol = 0;        // start of leading spaces
+  colnr_T leading_space_end = 0;  // end of contiguous initial spaces
 
   bool lcs_eol_todo = true;  // need to keep track of this even if lcs_eol is NUL
   const schar_T lcs_eol = wp->w_p_lcs_chars.eol;  // 'eol' value
@@ -1477,6 +1478,17 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, b
     }
   }
 
+  if (wp->w_buffer->b_p_lsw > 0 && !has_foldtext && draw_text) {
+    while (ptr[leading_space_end] == ' ') {
+      leading_space_end++;
+    }
+    if (ptr[leading_space_end] == NUL) {
+      leading_space_end = 0;
+    } else {
+      extra_check = true;
+    }
+  }
+
   // 'nowrap' or 'wrap' and a single line that doesn't fit: Advance to the
   // first character to be displayed.
   const int start_vcol = wp->w_p_wrap
@@ -2413,6 +2425,15 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, b
           saved_attr2 = wlv.char_attr;  // save current attr
           mb_c = schar_get_first_codepoint(mb_schar);
         }
+
+        if (wp->w_buffer->b_p_lsw > 0
+            && leading_space_end > 0
+            && prev_ptr < line + leading_space_end
+            && prev_ptr[0] == ' ') {
+          wlv.n_extra = (int)wp->w_buffer->b_p_lsw - 1;
+          wlv.sc_extra = mb_schar;
+          wlv.sc_final = NUL;
+        }
       }
 
       // Handling of non-printable characters.
diff --git a/src/nvim/indent.c b/src/nvim/indent.c
index 806526b734..1f027a143b 100644
--- a/src/nvim/indent.c
+++ b/src/nvim/indent.c
@@ -371,19 +371,19 @@ int get_sts_value(void)
 /// Count the size (in window cells) of the indent in the current line.
 int get_indent(void)
 {
-  return indent_size_ts(get_cursor_line_ptr(), curbuf->b_p_ts, curbuf->b_p_vts_array);
+  return indent_size_ts(get_cursor_line_ptr(), curbuf->b_p_ts, curbuf->b_p_vts_array, 0);
 }
 
 /// Count the size (in window cells) of the indent in line "lnum".
 int get_indent_lnum(linenr_T lnum)
 {
-  return indent_size_ts(ml_get(lnum), curbuf->b_p_ts, curbuf->b_p_vts_array);
+  return indent_size_ts(ml_get(lnum), curbuf->b_p_ts, curbuf->b_p_vts_array, 0);
 }
 
 /// Count the size (in window cells) of the indent in line "lnum" of buffer "buf".
 int get_indent_buf(buf_T *buf, linenr_T lnum)
 {
-  return indent_size_ts(ml_get_buf(buf, lnum), buf->b_p_ts, buf->b_p_vts_array);
+  return indent_size_ts(ml_get_buf(buf, lnum), buf->b_p_ts, buf->b_p_vts_array, 0);
 }
 
 /// Compute the size of the indent (in window cells) in line "ptr",
@@ -407,12 +407,14 @@ int indent_size_no_ts(char const *ptr)
 }
 
 /// Compute the size of the indent (in window cells) in line "ptr",
-/// using tabstops
-int indent_size_ts(char const *ptr, OptInt ts, colnr_T *vts)
+/// using tabstops. When "lsw" > 0, each leading space counts as "lsw"
+/// cells instead of 1.
+int indent_size_ts(char const *ptr, OptInt ts, colnr_T *vts, int lsw)
   FUNC_ATTR_NONNULL_ARG(1) FUNC_ATTR_PURE
 {
   assert(char2cells(' ') == 1);
 
+  int const space_width = (lsw > 0) ? lsw : 1;
   int vcol = 0;
   int tabstop_width, next_tab_vcol;
 
@@ -432,13 +434,13 @@ int indent_size_ts(char const *ptr, OptInt ts, colnr_T *vts)
       do {
         char const c = *ptr++;
         if (c == ' ') {
-          cur_vcol++;
+          cur_vcol += space_width;
         } else if (c == TAB) {
           break;
         } else {
           return cur_vcol;
         }
-      } while (cur_vcol != vcol);
+      } while (cur_vcol < vcol);
     }
 
     tabstop_width = *last_tabstop;
@@ -449,8 +451,8 @@ int indent_size_ts(char const *ptr, OptInt ts, colnr_T *vts)
   while (true) {
     char const c = *ptr++;
     if (c == ' ') {
-      vcol++;
-      next_tab_vcol += (vcol == next_tab_vcol) ? tabstop_width : 0;
+      vcol += space_width;
+      next_tab_vcol += (vcol >= next_tab_vcol) ? tabstop_width : 0;
     } else if (c == TAB) {
       vcol = next_tab_vcol;
       next_tab_vcol += tabstop_width;
@@ -850,6 +852,7 @@ int get_breakindent_win(win_T *wp, char *line)
   static bool prev_no_ts = false;  // cached no_ts value
   static unsigned prev_dy_uhex = 0;   // cached 'display' "uhex" value
   static char *prev_flp = NULL;  // cached formatlistpat value
+  static int prev_lsw = 0;  // cached 'leadingspacewidth' value
   int bri = 0;
   // window width minus window margin space, i.e. what rests for text
   const int eff_wwidth = wp->w_view_width - win_col_off(wp) + win_col_off2(wp);
@@ -870,6 +873,7 @@ int get_breakindent_win(win_T *wp, char *line)
   if (prev_fnum != wp->w_buffer->b_fnum
       || prev_ts != wp->w_buffer->b_p_ts
       || prev_vts != wp->w_buffer->b_p_vts_array
+      || prev_lsw != (int)wp->w_buffer->b_p_lsw
       || prev_tick != buf_get_changedtick(wp->w_buffer)
       || prev_listopt != wp->w_briopt_list
       || prev_no_ts != no_ts
@@ -882,12 +886,14 @@ int get_breakindent_win(win_T *wp, char *line)
     prev_line = xstrdup(line);
     prev_ts = wp->w_buffer->b_p_ts;
     prev_vts = wp->w_buffer->b_p_vts_array;
+    prev_lsw = (int)wp->w_buffer->b_p_lsw;
     if (wp->w_briopt_vcol == 0) {
       if (no_ts) {
         prev_indent = indent_size_no_ts(line);
       } else {
         prev_indent = indent_size_ts(line, wp->w_buffer->b_p_ts,
-                                     wp->w_buffer->b_p_vts_array);
+                                     wp->w_buffer->b_p_vts_array,
+                                     (int)wp->w_buffer->b_p_lsw);
       }
     }
     prev_tick = buf_get_changedtick(wp->w_buffer);
diff --git a/src/nvim/option.c b/src/nvim/option.c
index 4eeb6b6acf..2472367ffb 100644
--- a/src/nvim/option.c
+++ b/src/nvim/option.c
@@ -3043,6 +3043,7 @@ static const char *validate_num_option(OptIndex opt_idx, OptInt *newval, char *e
   case kOptUpdatetime:
   case kOptSidescroll:
   case kOptFoldlevel:
+  case kOptLeadingspacewidth:
   case kOptShiftwidth:
   case kOptTextwidth:
   case kOptWritedelay:
@@ -4984,6 +4985,8 @@ void *get_varp_from(vimoption_T *p, buf_T *buf, win_T *win)
     return &(buf->b_p_qe);
   case kOptReadonly:
     return &(buf->b_p_ro);
+  case kOptLeadingspacewidth:
+    return &(buf->b_p_lsw);
   case kOptScrollback:
     return &(buf->b_p_scbk);
   case kOptSmartindent:
@@ -5319,6 +5322,8 @@ void buf_copy_options(buf_T *buf, int flags)
       buf->b_p_ai = p_ai;
       COPY_OPT_SCTX(buf, kBufOptAutoindent);
       buf->b_p_ai_nopaste = p_ai_nopaste;
+      buf->b_p_lsw = p_lsw;
+      COPY_OPT_SCTX(buf, kBufOptLeadingspacewidth);
       buf->b_p_sw = p_sw;
       COPY_OPT_SCTX(buf, kBufOptShiftwidth);
       buf->b_p_scbk = p_scbk;
diff --git a/src/nvim/option_vars.h b/src/nvim/option_vars.h
index a8f26f4730..b5a01da493 100644
--- a/src/nvim/option_vars.h
+++ b/src/nvim/option_vars.h
@@ -412,6 +412,7 @@ EXTERN char *p_lispwords;       ///< 'lispwords'
 EXTERN OptInt p_ls;             ///< 'laststatus'
 EXTERN OptInt p_stal;           ///< 'showtabline'
 EXTERN char *p_lcs;             ///< 'listchars'
+EXTERN OptInt p_lsw;            ///< 'leadingspacewidth'
 EXTERN int p_lz;                ///< 'lazyredraw'
 EXTERN int p_lpl;               ///< 'loadplugins'
 EXTERN int p_magic;             ///< 'magic'
diff --git a/src/nvim/options.lua b/src/nvim/options.lua
index 7cb8f7f443..2f8d4700ec 100644
--- a/src/nvim/options.lua
+++ b/src/nvim/options.lua
@@ -5252,6 +5252,29 @@ local options = {
       type = 'boolean',
       varname = 'p_lz',
     },
+    {
+      abbreviation = 'lsw',
+      defaults = 0,
+      desc = [=[
+        Visual width (in screen columns) to use when rendering each leading
+        space character.  When set to 0 (the default), leading spaces render
+        at their normal width of 1 column.  When set to a value > 0, each
+        leading space occupies that many columns on screen, without changing
+        the buffer content.  Tab characters in leading whitespace are not
+        affected (use 'tabstop' for those).
+
+        This is useful when working with codebases that use narrow space
+        indentation (e.g., 2-space) but you prefer a wider visual indent.
+        Example: with 2-space indentation and 'leadingspacewidth' set to 2,
+        each indent level displays as 4 columns wide.
+      ]=],
+      full_name = 'leadingspacewidth',
+      redraw = { 'current_buffer' },
+      scope = { 'buf' },
+      short_desc = N_('visual width of leading space characters'),
+      type = 'number',
+      varname = 'p_lsw',
+    },
     {
       abbreviation = 'lhi',
       cb = 'did_set_xhistory',
diff --git a/src/nvim/plines.c b/src/nvim/plines.c
index 1fafdf2206..aaef82c1a5 100644
--- a/src/nvim/plines.c
+++ b/src/nvim/plines.c
@@ -102,6 +102,17 @@ CSType init_charsize_arg(CharsizeArg *csarg, win_T *wp, linenr_T lnum, char *lin
   csarg->virt_row = -1;
   csarg->indent_width = INT_MIN;
   csarg->use_tabstop = !wp->w_p_list || wp->w_p_lcs_chars.tab1;
+  csarg->leading_space_width = (int)wp->w_buffer->b_p_lsw;
+  csarg->leading_space_end = 0;
+  if (csarg->leading_space_width > 0) {
+    char *p = line;
+    while (*p == ' ') {
+      p++;
+    }
+    if (*p != NUL) {
+      csarg->leading_space_end = (colnr_T)(p - line);
+    }
+  }
 
   if (lnum > 0) {
     if (marktree_itr_get_filter(wp->w_buffer->b_marktree, lnum - 1, 0, lnum, 0,
@@ -111,6 +122,7 @@ CSType init_charsize_arg(CharsizeArg *csarg, win_T *wp, linenr_T lnum, char *lin
   }
 
   if (csarg->virt_row >= 0
+      || csarg->leading_space_end > 0
       || (wp->w_p_wrap && (wp->w_p_lbr || wp->w_p_bri || *get_showbreak_value(wp) != NUL))) {
     return kCharsizeRegular;
   } else {
@@ -153,6 +165,10 @@ CharSize charsize_regular(CharsizeArg *csarg, char *const cur, colnr_T const vco
     size = kInvalidByteCells;
   } else {
     size = ptr2cells(cur);
+    if (cur_char == ' ' && csarg->leading_space_width > 0
+        && cur >= line && cur < line + csarg->leading_space_end) {
+      size = csarg->leading_space_width;
+    }
     is_doublewidth = size == 2 && cur_char >= 0x80;
   }
 
diff --git a/src/nvim/plines.h b/src/nvim/plines.h
index 5ed5ef32e0..d3aa484177 100644
--- a/src/nvim/plines.h
+++ b/src/nvim/plines.h
@@ -28,6 +28,8 @@ typedef struct {
   int cur_text_width_right;  ///< Width of virtual text right of cursor.
 
   int max_head_vcol;         ///< See charsize_regular().
+  colnr_T leading_space_end; ///< Byte offset where leading spaces end (0 if disabled).
+  int leading_space_width;   ///< Value of 'leadingspacewidth' (0 = disabled).
   MarkTreeIter iter[1];
 } CharsizeArg;
 
diff --git a/test/unit/indent_spec.lua b/test/unit/indent_spec.lua
index a1c3dc4a75..d0af9749f2 100644
--- a/test/unit/indent_spec.lua
+++ b/test/unit/indent_spec.lua
@@ -34,18 +34,18 @@ end)
 describe('indent_size_ts()', function()
   itp('works for spaces', function()
     local line = to_cstr((' '):rep(7) .. 'a ')
-    eq(7, indent.indent_size_ts(line, 100, nil))
+    eq(7, indent.indent_size_ts(line, 100, nil, 0))
   end)
 
   itp('works for tabs and spaces', function()
     local line = to_cstr('   \t  \t \t\t   a ')
-    eq(19, indent.indent_size_ts(line, 4, nil))
+    eq(19, indent.indent_size_ts(line, 4, nil, 0))
   end)
 
   itp('works for tabs and spaces with empty vts', function()
     local vts = ffi.new('int[1]') -- zero initialized => first element (size) == 0
     local line = to_cstr('   \t  \t \t\t       a ')
-    eq(23, indent.indent_size_ts(line, 4, vts))
+    eq(23, indent.indent_size_ts(line, 4, vts, 0))
   end)
 
   itp('works for tabs and spaces with vts', function()
@@ -55,6 +55,11 @@ describe('indent_size_ts()', function()
     vts[2] = 2
 
     local line = to_cstr('      \t  \t \t\t   a ')
-    eq(18, indent.indent_size_ts(line, 4, vts))
+    eq(18, indent.indent_size_ts(line, 4, vts, 0))
+  end)
+
+  itp('accounts for leadingspacewidth', function()
+    local line = to_cstr('  \ta')
+    eq(8, indent.indent_size_ts(line, 4, nil, 2))
   end)
 end)
diff --git a/test/functional/ui/leadingspacewidth_spec.lua b/test/functional/ui/leadingspacewidth_spec.lua
new file mode 100644
index 0000000000..4ca1772c43
--- /dev/null
+++ b/test/functional/ui/leadingspacewidth_spec.lua
@@ -0,0 +1,180 @@
+local t = require('test.testutil')
+local n = require('test.functional.testnvim')()
+local Screen = require('test.functional.ui.screen')
+
+local clear = n.clear
+local feed = n.feed
+local command = n.command
+local api = n.api
+local eq = t.eq
+local fn = n.fn
+
+--- @param lines string[]
+local function set_lines(lines)
+  api.nvim_buf_set_lines(0, 0, -1, true, lines)
+end
+
+describe("'leadingspacewidth'", function()
+  local screen
+
+  before_each(function()
+    clear()
+    screen = Screen.new(40, 8)
+  end)
+
+  describe('option', function()
+    it('defaults to 0', function()
+      eq(0, api.nvim_get_option_value('leadingspacewidth', { buf = 0 }))
+    end)
+
+    it('can be set and get via abbreviation', function()
+      command('set lsw=3')
+      eq(3, api.nvim_get_option_value('leadingspacewidth', { buf = 0 }))
+    end)
+
+    it('rejects negative values', function()
+      local ok = pcall(command, 'set lsw=-1')
+      eq(false, ok)
+    end)
+
+    it('is buffer-local', function()
+      command('setlocal lsw=2')
+      command('new')
+      eq(0, api.nvim_get_option_value('leadingspacewidth', { buf = 0 }))
+    end)
+  end)
+
+  describe('display', function()
+    it('expands leading spaces when lsw=2', function()
+      set_lines({ '  hello', '    world' })
+      command('set lsw=2')
+      screen:expect([[
+^    hello                               |
+        world                           |
+{1:~                                       }|*5
+                                        |
+]])
+    end)
+
+    it('expands leading spaces when lsw=3', function()
+      set_lines({ '  hello', '    world' })
+      command('set lsw=3')
+      screen:expect([[
+^      hello                             |
+            world                       |
+{1:~                                       }|*5
+                                        |
+]])
+    end)
+
+    it('does not affect non-leading spaces', function()
+      set_lines({ '  hello world' })
+      command('set lsw=2')
+      screen:expect([[
+^    hello world                         |
+{1:~                                       }|*6
+                                        |
+]])
+    end)
+
+    it('does not affect all-whitespace lines', function()
+      set_lines({ '  hello', '   ', '  world' })
+      command('set lsw=2')
+      screen:expect([[
+^    hello                               |
+                                        |
+    world                               |
+{1:~                                       }|*4
+                                        |
+]])
+    end)
+
+    it('does not affect tab characters', function()
+      set_lines({ '\thello' })
+      command('set tabstop=4 lsw=2')
+      screen:expect([[
+   ^ hello                               |
+{1:~                                       }|*6
+                                        |
+]])
+    end)
+
+    it('lsw=0 disables expansion (default)', function()
+      set_lines({ '  hello' })
+      command('set lsw=0')
+      screen:expect([[
+^  hello                                 |
+{1:~                                       }|*6
+                                        |
+]])
+    end)
+
+    it('lsw=1 renders at normal width', function()
+      set_lines({ '  hello' })
+      command('set lsw=1')
+      screen:expect([[
+^  hello                                 |
+{1:~                                       }|*6
+                                        |
+]])
+    end)
+
+    it('cursor position is correct on leading spaces', function()
+      set_lines({ '  hello' })
+      command('set lsw=2')
+      feed('0')
+      eq(1, fn.col('.'))
+    end)
+
+    it('works with multiple indent levels', function()
+      set_lines({ 'no indent', '  level 1', '    level 2', '      level 3' })
+      command('set lsw=2')
+      screen:expect([[
+^no indent                               |
+    level 1                             |
+        level 2                         |
+            level 3                     |
+{1:~                                       }|*3
+                                        |
+]])
+    end)
+  end)
+
+  describe('with listchars', function()
+    it('composes with lead listchar', function()
+      set_lines({ '  hello' })
+      command('set lsw=2 list listchars=lead:.')
+      screen:expect([[
+{1:^.}.{1:.}.hello                               |
+{1:~                                       }|*6
+                                        |
+]])
+    end)
+
+    it('composes with leadmultispace listchar', function()
+      set_lines({ '    hello' })
+      command('set lsw=2 list listchars=leadmultispace:.-')
+      screen:expect([[
+{1:^.}.{1:-}-{1:.}.{1:-}-hello                           |
+{1:~                                       }|*6
+                                        |
+]])
+    end)
+  end)
+
+  describe('cursor navigation', function()
+    it('virtcol returns correct virtual column', function()
+      set_lines({ '  hello' })
+      command('set lsw=2')
+      feed('0')
+      eq(2, fn.virtcol('.'))
+    end)
+
+    it('virtcol of text after indent is correct', function()
+      set_lines({ '  hello' })
+      command('set lsw=2')
+      feed('0w')
+      eq(5, fn.virtcol('.'))
+    end)
+  end)
+end)
